### ISSUE TITLE
macOS x86_64 requires frame pointers to unwind functions with huge stack frames

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#74640]
Currently macOS x86_64 forcibly has frame pointers enabled, but this is an unfortunate performance loss relative to other platforms. There's no fundamental need to use frame pointers on macOS and it'd be a nice perf boost if we could stop requiring them. https://github.com/rust-lang/rust/pull/74575 was a recent attempt to disable frame pointers on macOS but it ran into a lot of segfaults in the compiler. Some debugging showed that this program will also segfault:

```rust
fn main() {
    let _huge_stack = [0u64; 1024];
    panic!("x");
}
```

run with:

```sh
$ rustc main.rs -C force-frame-pointers=no
$ ./main
thread 'main' panicked at 'x', main.rs:3:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
zsh: segmentation fault  ./main
```

The issue appears to be with how the unwind information for the `main` function is encoded. macOS uses "compact unwind information" which basically means "not dwarf". I believe that this is an LLVM bug where the compact unwind information for `main` is incorrect when stack probes are also used. The `main` function requires a stack probe, and its unwind information is encoded in the `UNWIND_X86_64_MODE_STACK_IND` mode which is [documented as](https://gist.github.com/alexcrichton/42a2d7bc175d408a2a9682629425f7ef):

```
// UNWIND_X86_64_MODE_STACK_IND:
//    A "frameless" (RBP not used as frame pointer) function large constant 
//    stack size.  This case is like the previous, except the stack size is too
//    large to encode in the compact unwind encoding.  Instead it requires that 
//    the function contains "subq $nnnnnnnn,RSP" in its prolog.  The compact 
//    encoding contains the offset to the nnnnnnnn value in the function in
//    UNWIND_X86_64_FRAMELESS_STACK_SIZE.  
```

That all matches the description of `main` (doesn't use RBP, big stack that it probably can't fit in an otherwise compact encoding), but the `main` function doesn't have a subtraction to set the stack pointer. Instead the main function calls `__rust_probestack` and then executes `subq $rax, $rsp` where `$rax` has the stack size.

Basically I think this boils down to the unwind information generated by LLVM being incorrect for functions which trigger usage of `__rust_probestack` (stack probe guards). I don't think there's much we can do to fix this in rust-lang/rust but I wanted to file an issue to track this.

**Comment by eggyal** [comment]
Could #74771 be the issue underlying this?

**Comment by mstange** [comment]
Having framepointers is a good default, because it enables simple and reliable stack unwinding for profilers and crash reporters. On macOS, Xcode and clang leave framepointers enabled by default. All macOS system libraries ship with framepointers.
If this issue ever gets fixed, I'd like to argue that the default should still remain the same as it is today: framepointers should be present by default.

**Comment by workingjubilee** [comment]
For aarch64-apple-darwin, frame pointers are mandated as part of the ABI, with only leaf functions and tail calls being able to back out of this.
